### PR TITLE
[Generator] Support turn-level rewards in SkyRLGymGenerator

### PR DESCRIPTION
**Pull request by tyler-griggs**
Addressing issue #201 

## What does this PR do?
Add support for turn-level rewards in SkyRLGymGenerator. The trainer already supports token-level rewards, but previously the SkyRLGymGenerator ignored rewards returned by `step()` except for the final turn. This PR tracks per-`step` rewards and builds a token-level reward list that associates each `step()` reward with the final token in the assistant's response that led to the reward. 

## Tests
Added CPU tests for main functionality and edge cases (e.g., response truncation due to exceeding max length)

e2e test with step-level rewards: **WIP**

### CODE REVIEW

File: skyrl-train/examples/turn_level_rewards/README.md
Comment (SumanthRH) [inline skyrl-train/examples/turn_level_rewards/README.md:1]: to be filled?

File: skyrl-train/skyrl_train/generators/skyrl_gym_generator.py
Comment (SumanthRH) [inline skyrl-train/skyrl_train/generators/skyrl_gym_generator.py:1]: Should the changes be in the same generator or do we just have a custom generator example for step level rewards? Latter seemed like it's easier to maintain (this is also a niche example at the moment)

Comment (tyler-griggs) [reply skyrl-train/skyrl_train/generators/skyrl_gym_generator.py:1]: Hm, I'm partial to keeping it as one generator. I actually think the custom generator example will be harder to maintain. At least, I have a hunch it will quickly become far out of date as we iterate on the skyrl gym generator. I don't actually think this is a very intrusive change, just one list with post-processing. We are currently ignoring most of the returns of `step()`, which we probably shouldn't be doing anyways. 

That said, I do feel that the SkyRL Gym Generator is getting (too) complex, but I think the way to do it is by pruning some of the code paths like `retokenize_chat_history`.

File: skyrl-train/examples/turn_level_rewards/README.md
Comment (tyler-griggs) [reply skyrl-train/examples/turn_level_rewards/README.md:1]: Yeah it's a WIP

File: skyrl-train/skyrl_train/generators/skyrl_gym_generator.py
Comment (SumanthRH) [reply skyrl-train/skyrl_train/generators/skyrl_gym_generator.py:1]: >  I actually think the custom generator example will be harder to maintain. At least, I have a hunch it will quickly become far out of date as we iterate on the skyrl gym generator.

How so? We actually don't have to support all the features we will add to skyrl gym generator in the future. Like step level rewards doesn't have to support `use_conversation_multi_turn`

 I guess we might change the stuff around chat templating and tokenization soon, but a standalone generator should still work okay?
 
 > I do feel that the SkyRL Gym Generator is getting (too) complex, but I think the way to do it is by pruning some of the code paths like retokenize_chat_history.
 
 Hmm yeah this is my main concern: SkyRLGymGenerator is **the** example for how a generator should look like with SkyRL, and it's getting pretty involved (ex: the `use_conversation_multi_turn` codepaths, etc)
 
 
 I'm not too strongly attached to either btw, up to you

Comment (tyler-griggs) [reply skyrl-train/skyrl_train/generators/skyrl_gym_generator.py:1]: > I guess we might change the stuff around chat templating and tokenization soon, but a standalone generator should still work okay?
Yeah, this was one pending set of changes I was referring to. And I do agree that the standalone generator would probably work, it just wouldn't be all that useful if it falls out of date with what's supported in the SkyRLGymGenerator. 

Ultimately, the reason I lean towards keeping this in is a combination of the change being relatively small, and I do think we should be making use of `step()` outputs instead of ignoring them. 

Charlie and I spoke recently about ways to simplify the SkyRL Gym Generator and prune some of the bloat, and I think we can shed it that way.

**Comment by tyler-griggs** [comment]
Okay now ready for review! I'll add a wandb screenshot shortly.

File: skyrl-train/examples/turn_level_rewards/gsm8k_multi_turn_dataset.py [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+    args.output_dir = os.path.expanduser(args.output_dir)
+
+    data_source = "openai/gsm8k"
+
+    dataset = datasets.load_dataset(data_source, "main")
+
+    train_dataset = dataset["train"]
+    val_dataset = dataset["test"]
+
+    instruction_following = 'Let\'s think step by step and output the final answer after "####".'
```
Comment (SumanthRH) [inline skyrl-train/examples/turn_level_rewards/gsm8k_multi_turn_dataset.py]: we can assume max_turns is > 1 for this example and thus the prompt can be :

```suggestion
    instruction_following = 'Let\'s think step by step and output a tentative numeric answer after "####".'
```

Comment (tyler-griggs) [reply skyrl-train/examples/turn_level_rewards/gsm8k_multi_turn_dataset.py]: Good call, thanks
